setup_unit_tests.py

```diff
@@ -0,0 +1,488 @@
+#!/usr/bin/env python3
+# PYTHON_ARGCOMPLETE_OK
+
+"""
+Extract unit tests from the WDL spec (1.1+) and create a conformance file that is conformance to our representation for this test suite
+"""
+import hashlib
+import subprocess
+import sys
+
+import regex as re
+import os
+import glob
+
+import json
+from ruamel.yaml import YAML
+import argparse
+import argcomplete
+from pathlib import Path
+import re
+import shutil
+from typing import Optional, Union, List, Set, Callable, Any, Dict
+
+from lib import convert_type
+
+import WDL
+
+TEST_RE = re.compile(
+    r"^<details>\s*<summary>\s*Example: (.+?)\s*```wdl(.+?)```\s*</summary>\s*(?:<p>\s*(?:Example input:\s*```json(.*?)```)?\s*(?:Example output:\s*```json(.*?)```)?\s*(?:Test config:\s*```json(.*)```)?\s*</p>\s*)?</details>$",
+    re.I | re.S,
+)
+FILENAME_RE = re.compile(r"(.+?)(_fail)?(_task)?.wdl")
+VERSION_RE = re.compile(r"version ([\d.]+)")
+
+regex_output_section_str = r"(?<=output)(?:\s*{\s*)([\s\S]+?)(?:})"
+regex_var_types_str = r"([\w\[\]+?]+)\s(\w+)(?:[\s\S]*?(?= =))"
+
+
+def get_not_none(d: Optional[Dict[Any, Any]], k: str) -> Any:
+    # same as d.get(k) but return None if d is None
+    if d is not None:
+        return d.get(k)
+    return None
+
+def index_not_none(l: Optional[List[Any]], i: int) -> Any:
+    if l is not None:
+        return l[i]
+    return None
+
+def extract_output_types_defer(wdl_file):
+    """
+    The main issue with using miniwdl's parser is parsing the wdl file is dependent on miniwdl,
+    so if the spec changes and miniwdl is unable to support it, this test framework won't be able to run.
+    The simplest way around this (that isn't creating another parser, maybe pyparsing?) is probably something like this;
+    regex won't work as regular languages are too basic.
+
+    Thus use a simple iterator to detect all variable names instead of miniwdl's parser
+    To be used if a wdl file has a type that miniwdl doesn't support, ex Object
+    """
+    regex_var_types = re.compile(regex_var_types_str)
+
+    with open(wdl_file, "r") as f:
+        wdl_contents = f.read()
+
+    var_types = {}
+    # can't regex this as regex cannot handle nested constructs
```

Comment:
You say this but then proceed to somehow parse everything with a regex anyway?

Reply:
Yeah the comment is a little misleading. I think it was left from a previous implementation. What I meant is that using a regex expression on the entire WDL file (represented as a string) will not work as regex cannot handle nested constructs. Instead, I iterate line by line and basically grab everything left of the `=` sign.

I changed the comment so it makes more sense